include/oneapi/dpl/pstl/iterator_impl.h

```diff
@@ -157,6 +190,22 @@ namespace oneapi
 {
 namespace dpl
 {
+
+template <typename T>
+struct is_indirectly_device_accessible_iterator
+    : decltype(oneapi::dpl::__internal::__is_onedpl_indirectly_device_accessible_iterator(std::declval<T>()))
+{
+    static_assert(
+        std::is_same_v<decltype(decltype(oneapi::dpl::__internal::__is_onedpl_indirectly_device_accessible_iterator(
+                           std::declval<T>()))::value),
+                       const bool>,
+        "Return type of is_onedpl_indirectly_device_accessible_iterator does not have the characteristics of a "
+        "bool_constant");
+};
```

Comment:
Does it make sense to redefine it like
```suggestion
template <typename T>
using is_indirectly_device_accessible_iterator_t =
    std::integral_constant<bool, decltype(oneapi::dpl::__internal::__is_onedpl_indirectly_device_accessible_iterator(
                                     std::declval<T>()))::value>;
```

Reply:
As stated before, I disagree with the `_t` part, but as for an alias instead of a `struct`...

It removes our ability to `static_assert`.  After a discussion offline with @reble , I believe its best to force people into returning types with base characteristics of `bool_constant` rather than permitting `integral_constant` with `value` 
constant convertible to `bool`.

This allows the specification to be simple (and only discuss `bool_constant`). We don't have a compelling use case for needing to support `integral_constant` with `value` constant convertible to `bool`. With the `static_assert`, it provides a useful error which will correct any reliance on undefined behavior outside the simpler specification.

However, what you describe is a possibility, and if we have a compelling use case of an possible user type which wants to use an `integral_constant` directly I'm open to changing my mind here (which probably means changing the spec PR as well).

Reply:
Both versions force the public trait to be a `bool_constant`, regardless of what users return from the ADL based customization point, which I think is important.  This version in the PR with the static assertion requires explicitly that users return `bool_constant` from their ADL based customization point functions.  If we decide to switch to a more permissive specification, then your suggestion should be what we use for the implementation.